### PR TITLE
feat: Implement `optionRender`  API

### PR DESCRIPTION
**Pull request by RedJue**
[[中文版模板 / Chinese template](https://github.com/ant-design/ant-design/blob/master/.github/PULL_REQUEST_TEMPLATE/pr_cn.md?plain=1)]

### 🤔 This is a ...

- [x] New feature
- [ ] Bug fix
- [ ] Site / documentation update
- [ ] Demo update
- [ ] Component style update
- [ ] TypeScript definition update
- [ ] Bundle size optimization
- [ ] Performance optimization
- [ ] Enhancement feature
- [ ] Internationalization
- [ ] Refactoring
- [ ] Code style optimization
- [ ] Test Case
- [ ] Branch merge
- [ ] Workflow
- [ ] Other (about what?)

### 🔗 Related issue link
#45445

### 💡 Background and solution
image

### 📝 Changelog

| Language   | Changelog |
| ---------- | --------- |
| 🇺🇸 English |   Select support `optionRender`  API       |
| 🇨🇳 Chinese |    选择器支持  `optionRender` 方法      |

### ☑️ Self-Check before Merge

⚠️ Please check all items below before requesting a reviewing. ⚠️

- [ ] Doc is updated/provided or not needed
- [ ] Demo is updated/provided or not needed
- [ ] TypeScript definition is updated/provided or not needed
- [ ] Changelog is provided or not needed

---

### <samp>🤖 Generated by Copilot at a36076f</samp>

This pull request adds a new feature to the `Select` component that allows users to customize the rendering of the dropdown options with the `optionRender` prop. It updates the documentation, the dependency, and the demo code to showcase this feature.

### <samp>🤖 Generated by Copilot at a36076f</samp>

*  Add a new demo `option-render` to show how to customize the rendering of the dropdown options in the `Select` component ([link](https://github.com/ant-design/ant-design/pull/45529/files?diff=unified&w=0#diff-9b086b1f8c9c28a3006203bf1c52e9eff4a23c540e5ea10938b08e69ae9f48a1R1-R7), [link](https://github.com/ant-design/ant-design/pull/45529/files?diff=unified&w=0#diff-448234532d7d44c800e124294730c0980a333fc8bf04d2f2d72a2475549b82f6R1-R43), [link](https://github.com/ant-design/ant-design/pull/45529/files?diff=unified&w=0#diff-bf56ff1410a74a10dffb8c98ad2163749383eb3d25cc89fd10e747bbd8004452R44), [link](https://github.com/ant-design/ant-design/pull/45529/files?diff=unified&w=0#diff-07bf0b280d743be2023d98c096d6127d8fa2d5e9baa0c23ab8ee2a4f1c0db96bR45))
*  Add a new prop `optionRender` to the `Select` component to allow users to pass a function that returns a React node for each option ([link](https://github.com/ant-design/ant-design/pull/45529/files?diff=unified&w=0#diff-bf56ff1410a74a10dffb8c98ad2163749383eb3d25cc89fd10e747bbd8004452R89), [link](https://github.com/ant-design/ant-design/pull/45529/files?diff=unified&w=0#diff-07bf0b280d743be2023d98c096d6127d8fa2d5e9baa0c23ab8ee2a4f1c0db96bR90))
*  Update the dependency version of `rc-select` to `~14.10.0` to use the new prop `optionRender` ([link](https://github.com/ant-design/ant-design/pull/45529/files?diff=unified&w=0#diff-7ae45ad102eab3b6d7e7896acd08c427a9b25b346470d7bc6507b6481575d519L147-R147))
*  Remove a trailing whitespace in the `package.json` file ([link](https://github.com/ant-design/ant-design/pull/45529/files?diff=unified&w=0#diff-7ae45ad102eab3b6d7e7896acd08c427a9b25b346470d7bc6507b6481575d519L342-L341))

### CODE REVIEW

File: components/select/index.zh-CN.md [hunk 1]
```diff
@@ -86,6 +87,7 @@ demo:
 | optionFilterProp | 搜索时过滤对应的 `option` 属性，如设置为 `children` 表示对内嵌内容进行搜索。若通过 `options` 属性配置选项内容，建议设置 `optionFilterProp="label"` 来对内容进行搜索。 | string | `value` |  |
 | optionLabelProp | 回填到选择框的 Option 的属性值，默认是 Option 的子元素。比如在子元素需要高亮效果时，此值可以设为 `value`。[示例](https://codesandbox.io/s/antd-reproduction-template-tk678) | string | `children` |  |
 | options | 数据化配置选项内容，相比 jsx 定义会获得更好的渲染性能 | { label, value }\[] | - |  |
+| optionRender | 自定义渲染下拉选项 | (oriOption: FlattenOptionData\<BaseOptionType\> , info: { index: number }) => React.ReactNode | - | 5.11.0 |
```
Comment (yoyo837) [inline components/select/index.zh-CN.md:90]: 用这个属性对虚拟滚动兼容吗?

Comment (RedJue) [reply components/select/index.zh-CN.md:90]: 应该是兼容的，这个属性只对内容生效

File: components/select/demo/option-render.tsx [hunk 2]
```diff
@@ -0,0 +1,43 @@
...
+  return (
+    <Select
+      style={{ width: '100%' }}
+      options={options}
+      placeholder="Please select"
+      value={value}
+      onChange={setValue}
+      optionRender={(oriOption) => (
+        <Space>
+          <Radio checked={value === oriOption.value} />
```
Comment (afc163) [inline components/select/demo/option-render.tsx:35]: 用 Checkbox

File: components/select/demo/option-render.tsx [hunk 3]
```diff
@@ -0,0 +1,43 @@
...
+          <span>{oriOption.label}</span>
+        </Space>
+      )}
+    />
+  );
+};
+
+export default App;
```
[the remark this answers is not in the document]
Comment (RedJue) [reply components/select/demo/option-render.tsx:43]: 好👌

File: components/select/demo/option-render.tsx (hunk 2 above)
Comment (RedJue) [reply components/select/demo/option-render.tsx:35]: 本来用了 checkbox ，但想着这里是单选场景就换了 radio

File: components/select/demo/option-render.md [hunk 4]
```diff
@@ -0,0 +1,7 @@
+## zh-CN
+
+使用 `optionRender` 对下拉菜单项进行自由扩展。
+
+## en-US
+
+Use `optionRender` to freely expand dropdown menu items.
```
Comment (kiner-tang) [inline components/select/demo/option-render.md:7]:
```suggestion
Use `optionRender` to freely customize dropdown menu items.
```

File: components/select/demo/option-render.md [hunk 5]
```diff
@@ -0,0 +1,7 @@
+## zh-CN
+
+使用 `optionRender` 对下拉菜单项进行自由扩展。
```
Comment (kiner-tang) [inline components/select/demo/option-render.md:3]:
```suggestion
使用 `optionRender` 对下拉菜单项进行自定义。
```

File: components/select/index.en-US.md [hunk 6]
```diff
@@ -41,6 +41,7 @@ Select component to select value from options.
 <code src="./demo/big-data.tsx">Big Data</code>
 <code src="./demo/status.tsx">Status</code>
 <code src="./demo/placement.tsx">Placement</code>
+<code src="./demo/option-render.tsx">Customize the rendering dropdown options</code>
```
Comment (kiner-tang) [inline components/select/index.en-US.md:44]:
```suggestion
<code src="./demo/option-render.tsx" version=">=5.11.0">Customize the rendering dropdown options</code>
```

File: components/select/index.zh-CN.md [hunk 7]
```diff
@@ -42,6 +42,7 @@ demo:
 <code src="./demo/big-data.tsx">大数据</code>
 <code src="./demo/status.tsx">自定义状态</code>
 <code src="./demo/placement.tsx">弹出位置</code>
+<code src="./demo/option-render.tsx">自定义渲染下拉选项</code>
```
Comment (kiner-tang) [inline components/select/index.zh-CN.md:45]:
```suggestion
<code src="./demo/option-render.tsx" version=">=5.11.0">自定义渲染下拉选项</code>
```

File: components/select/index.en-US.md (hunk 6 above)
Comment (kiner-tang) [inline components/select/index.en-US.md:44]:
```suggestion
<code src="./demo/option-render.tsx">Customize the dropdown options</code>
```

Comment (RedJue) [reply components/select/index.en-US.md]: 这里意义都差不多，不改也没事

Comment (kiner-tang) [reply components/select/index.en-US.md]: 感觉不太像口语，不过本来就是文档，应该也可

**Review by kiner-tang — APPROVED** [review]

**Review by yoyo837 — APPROVED** [review]

File: components/select/index.zh-CN.md (hunk 1 above)
Comment (afc163) [inline components/select/index.zh-CN.md:90]: 别用缩写，oriOption => option 或 originalOption

Comment (RedJue) [reply components/select/index.zh-CN.md:90]: 1

File: components/select/demo/option-label-prop.tsx [hunk 8]
```diff
@@ -15,40 +41,16 @@ const App: React.FC = () => (
...
-      <Space>
-        <span role="img" aria-label="Japan">
-          🇯🇵
-        </span>
-        Japan (日本)
-      </Space>
-    </Option>
-    <Option value="korea" label="Korea">
+    options={options}
+    optionRender={(originalOption) => (
```
Comment (afc163) [inline components/select/demo/option-label-prop.tsx:45]: 看上去 option 就行了，好像也没有 original 和 current 的区别。

**Comment by davidhsing** [comment]
@afc163 大佬，请问这个 optionRender 功能，能加入到 4x 系列吗？
目前 antd v4.24.15 用的 rc-select ~14.1.18，这个版本的还没有加入 optionRender 选项。

**Comment by afc163** [comment]
4.x 目前不接受新特性了。

**Comment by wwq19930808** [comment]
Warning: React does not recognize the `optionRender` prop on a DOM element. If you intentionally want it to appear in the DOM as a custom attribute, spell it as lowercase `optionrender` instead. If you accidentally passed it from a parent component, remove it from the DOM element.
这个报错如何解决

**Pull request merged**